File: src/commands/connect.ts
import selector from "@inquirer/select";
import { Feed, FeedType, Project } from "@leaf/types";
import { makeRequest } from "@leaf/utils/rest";
import shell from "shelljs";
import {
  amqp,
  awsECS,
  awsS3,
  awsSQS,
  docker,
  file,
  kubernetes,
  redis,
  syslog,
} from "./sources";
import ora from "ora";

const SUPPORTED_SOURCES = {
  docker_logs: { name: "Docker Logs", handler: docker },
  kubernetes_logs: { name: "Kubernetes Logs", handler: kubernetes },
  amqp: { name: "AMQP/RabbitMQ", handler: amqp },
  redis: { name: "Redis", handler: redis },
  aws_ecs_metrics: { name: "AWS ECS Metrics", handler: awsECS },
  aws_s3: { name: "AWS S3", handler: awsS3 },
  aws_sqs: { name: "AWS SQS", handler: awsSQS },
  file: { name: "Log File", handler: file },
  syslog: { name: "Syslog", handler: syslog },
} as const;

export default async function () {
  try {
    const vectorInstalled = shell.exec("vector --help", {
      silent: true,
    });

    if (vectorInstalled.stderr) {
      throw new Error("vector doesn't exist");
    }

    const fetchingProjects = ora("Fetching Projects").start();
    const { success, data, error } = await makeRequest<{ projects: Project[] }>(
      "GET",
      "/users/@me"
    );

    if (!success || !data) {
      fetchingProjects.fail(
        `Failed to fetch projects: ${error?.message ?? "Unknown error"}`
      );

      return;
    }

    fetchingProjects.succeed("Fetched Projects");

    const project = await selector({
      message: "Select Project",
      choices: data.projects.map((p) => ({
        name: p.name,
        value: p.namespace,
        description: `lawg.dev/${p.namespace}`,
      })),
    });

    const selectedProject = data.projects.find(
      (p) => p.namespace === project
    ) as Project;

    const selectableFeeds = selectedProject.feeds
      .filter((f) => f.type === FeedType.APPLICATION)
      .map((f) => ({
        name: f.name,
        value: f.name,
      }));

    if (selectableFeeds.length === 0) {
      console.log(
        "No application feeds found, please create one in our dashboard."
      );
      return;
    }

    const feed = await selector({
      message: "Select An Application Feed",
      choices: selectableFeeds,
    });

    const selectedFeed = selectedProject.feeds.find(
      (f) => f.name === feed
    ) as Feed;

    const source = (await selector({
      message: "Select A Source",
      choices: Object.entries(SUPPORTED_SOURCES).map(([k, v]) => ({
        name: v.name,
        value: k,
      })),
    })) as keyof typeof SUPPORTED_SOURCES;

    SUPPORTED_SOURCES[source].handler(selectedFeed, selectedProject);
  } catch {
    console.log(
      "Vector does not exist, please install at https://vector.dev/docs/setup/installation/"
    );
  }
}

File: commands/connect.go
package commands

import (
	"leaf/commands/sources"
	"leaf/utils"
	"os/exec"

	"github.com/AlecAivazis/survey/v2"
	"github.com/urfave/cli/v2"
)

type Source struct {
	Name    string
	Handler func(feed utils.Feed, project utils.Project) error
}

var supportedSources = map[string]Source{
	"docker_logs": {
		Name:    "Docker Logs",
		Handler: sources.DockerLogs,
	},
	"kubernetes_logs": {
		Name:    "Kubernetes Logs",
		Handler: sources.KubernetesLogs,
	},
	"amqp": {
		Name:    "AMQP",
		Handler: sources.AMQP,
	},
	"redis": {
		Name:    "Redis",
		Handler: sources.Redis,
	},
	"aws_ecs_metrics": {
		Name:    "AWS ECS Metrics",
		Handler: sources.AwsECSMetrics,
	},
	"aws_s3": {
		Name:    "AWS S3",
		Handler: sources.AwsS3,
	},
	"file": {
		Name:    "File",
		Handler: sources.File,
	},
	"syslog": {
		Name:    "Syslog",
		Handler: sources.Syslog,
	},
}

var supportSourcesOrder = []string{
	"docker_logs",
	"kubernetes_logs",
	"amqp",
	"redis",
	"aws_ecs_metrics",
	"aws_s3",
	"file",
	"syslog",
}

func Connect(ctx *cli.Context) error {
	var cmd = exec.Command("vector", "--help")

	_, err := cmd.Output()
	if err != nil {
		return utils.ParsedError(err, "Unable to find Vector! Please install it at: https://vector.dev/docs/setup/installation")
	}

	state, err := utils.GetState()
	if err != nil {
		return utils.ParsedError(err, "Unable to get state", true)
	}

	projectSpinner := utils.Spinner.AddSpinner("Fetching Projects")
	utils.Spinner.Start()

	me, err := utils.GetMe(state.Token)

	if err != nil {
		projectSpinner.Error()
		projectSpinner.UpdateMessage(err.Error())
		utils.Spinner.Stop()

		return cli.Exit("", 1)
	}

	projectSpinner.UpdateMessage("Fetched Projects")
	projectSpinner.Complete()
	utils.Spinner.Stop()

	var projects = []string{}
	for _, project := range me.Data.Projects {
		projects = append(projects, project.Namespace)
	}

	var selectedProjectNamespace = ""
	survey.AskOne(&survey.Select{
		Message: "Select Project",
		Options: projects,
	}, &selectedProjectNamespace, survey.WithValidator(survey.Required))

	// get the selected project by namespace
	var selectedProject utils.Project
	for _, project := range me.Data.Projects {
		if project.Namespace == selectedProjectNamespace {
			selectedProject = project
		}
	}

	var feeds = []string{}
	for _, feed := range selectedProject.Feeds {
		// if feed is an event feed and not an application feed
		if feed.Type == 0 {
			continue
		}
		feeds = append(feeds, feed.Name)
	}

	if len(feeds) == 0 {
		println("No application feeds found")
		return nil
	}

	survey.AskOne(&survey.Select{
		Message: "Select An Application Feed",
		Options: feeds,
	}, &selectedProjectNamespace, survey.WithValidator(survey.Required))

	var selectedFeed utils.Feed
	for _, feed := range selectedProject.Feeds {
		if feed.Name == selectedProjectNamespace {
			selectedFeed = feed
		}
	}

	var sourceNames = []string{}
	for _, k := range supportSourcesOrder {
		sourceNames = append(sourceNames, supportedSources[k].Name)
	}

	var selectedSourceName = ""
	survey.AskOne(&survey.Select{
		Message: "Select A Source",
		Options: sourceNames,
	}, &selectedSourceName, survey.WithValidator(survey.Required))

	var selectedSource Source
	for _, source := range supportedSources {
		if source.Name == selectedSourceName {
			selectedSource = source
		}
	}

	return selectedSource.Handler(selectedFeed, selectedProject)
}
